Roslyn does not recognize a strongly named assembly

Repro steps:
* Open the attached solution [StrongNameTest.sln](http://tinyurl.com/qfu3zsu) in Visual Studio (with Roslyn installed)
* Try to build.

__ACTUAL:__
error CS8002: Referenced assembly 'Signed, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' does not have a strong name.

__EXPECTED:__
No errors, build succeeds.

This solution builds successfully with native C# compiler.
The referenced assembly has a strong name as could be checked by sn.exe:
```
StrongNameTest>sn -vf Signed.dll

Microsoft (R) .NET Framework Strong Name Utility  Version 4.0.30319.33440
Copyright (c) Microsoft Corporation.  All rights reserved.

Assembly 'Signed.dll' is valid

StrongNameTest>sn -q -Tp Signed.dll
Public key (hash algorithm: sha1):
0024000004800000940000000602000000240000525341310004000001000100ddbb397625d47d
8a90acf6e8d5abd57ec93fe21b735960ff5cc57e8606cdc2669ca8b24aaf64e02c30b926d6660a
b877ab61a9b510998fc1d060524da1c8d0058369c1193f4ced751769e6a699a9ae7750d589d305
0563eb1458edee3a53dba06fa5b46e3e97793fac6b90a96155f86c03804a6835cc9cb82cd04941
45b318a4

Public key token is 306f32af5e021d46
```
Also, it could be successfully installed into GAC:
```
StrongNameTest>gacutil /nologo /i Signed.dll
Assembly successfully added to the cache

StrongNameTest>gacutil /nologo /l Signed
The Global Assembly Cache contains the following assemblies:
  Signed, Version=1.0.0.0, Culture=neutral, PublicKeyToken=306f32af5e021d46, processorArchitecture=MSIL
```

<!---
@huboard:{"order":3.984375}
-->

--- a/src/Compilers/CSharp/Test/Emit/Attributes/InternalsVisibleToAndStrongNameTests.cs
+++ b/src/Compilers/CSharp/Test/Emit/Attributes/InternalsVisibleToAndStrongNameTests.cs
@@ -1650,8 +1650,8 @@ public class C
             }
         }
 
-        [Fact(Skip = "Bug 399")]
-        [WorkItem(399, "GitHub")]
+        [Fact]
+        [WorkItem(399, "https://github.com/dotnet/roslyn/issues/399")]
         public void Bug399()
         {
             // The referenced assembly Signed.dll from the repro steps

--- a/src/Compilers/Core/Portable/MetadataReader/PEModule.cs
+++ b/src/Compilers/Core/Portable/MetadataReader/PEModule.cs
@@ -474,32 +474,49 @@ namespace Microsoft.CodeAnalysis
                 throw new BadImageFormatException(string.Format(CodeAnalysisResources.InvalidAssemblyName, nameStr));
             }
 
-            string cultureName = (!culture.IsNil) ? reader.GetString(culture) : null;
+            string cultureName = culture.IsNil ? null : reader.GetString(culture);
             if (cultureName != null && !MetadataHelpers.IsValidMetadataIdentifier(cultureName))
             {
                 throw new BadImageFormatException(string.Format(CodeAnalysisResources.InvalidCultureName, cultureName));
             }
 
-            var hasPublicKey = (flags & AssemblyFlags.PublicKey) != 0;
-            var publicKeyOrToken = !publicKey.IsNil ? reader.GetBlobBytes(publicKey).AsImmutableOrNull() : default(ImmutableArray<byte>);
-            if (hasPublicKey)
+            ImmutableArray<byte> publicKeyOrToken = reader.GetBlobContent(publicKey);
+            bool hasPublicKey;
+
+            if (isReference)
             {
-                if (!MetadataHelpers.IsValidPublicKey(publicKeyOrToken))
+                hasPublicKey = (flags & AssemblyFlags.PublicKey) != 0;
+                if (hasPublicKey)
                 {
-                    throw new BadImageFormatException(CodeAnalysisResources.InvalidPublicKey);
+                    if (!MetadataHelpers.IsValidPublicKey(publicKeyOrToken))
+                    {
+                        throw new BadImageFormatException(CodeAnalysisResources.InvalidPublicKey);
+                    }
                 }
-            }
-            else if (isReference)
-            {
-                if (!publicKeyOrToken.IsDefaultOrEmpty && publicKeyOrToken.Length != AssemblyIdentity.PublicKeyTokenSize)
+                else
                 {
-                    throw new BadImageFormatException(CodeAnalysisResources.InvalidPublicKeyToken);
+                    if (!publicKeyOrToken.IsEmpty &&
+                        publicKeyOrToken.Length != AssemblyIdentity.PublicKeyTokenSize)
+                    {
+                        throw new BadImageFormatException(CodeAnalysisResources.InvalidPublicKeyToken);
+                    }
                 }
             }
             else
             {
-                // Assembly definitions do not contain public key tokens, but they may contain public key
-                // data without being marked as strong name signed (e.g. delay-signed assemblies).
+                // Assembly definitions never contain a public key token, they only can have a full key or nothing,
+                // so the flag AssemblyFlags.PublicKey does not make sense for them and is ignored.
+                // See Ecma-335, Partition II Metadata, 22.2 "Assembly : 0x20".
+                // This also corresponds to the behavior of the native C# compiler and sn.exe tool.
+                hasPublicKey = !publicKeyOrToken.IsEmpty;
+                if (hasPublicKey && !MetadataHelpers.IsValidPublicKey(publicKeyOrToken))
+                {
+                    throw new BadImageFormatException(CodeAnalysisResources.InvalidPublicKey);
+                }
+            }
+
+            if (publicKeyOrToken.IsEmpty)
+            {
                 publicKeyOrToken = default(ImmutableArray<byte>);
             }
 

--- a/src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/PseudoVariableTests.cs
+++ b/src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/PseudoVariableTests.cs
@@ -1050,8 +1050,8 @@ class C
             context.CompileExpression(
                 // typeof(Exception), typeof(A<B<object>>), typeof(B<A<object>[]>)
                 InspectionContextFactory.Empty.Add("$exception", "System.Exception, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089").
-                    Add("1", "A`1[[B`1[[System.Object, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], 397300B0-B, Version=1.2.2.2, Culture=neutral, PublicKeyToken=null]], 397300B0-A, Version=2.1.2.1, Culture=neutral, PublicKeyToken=null").
-                    Add("2", "B`1[[A`1[[System.Object, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]][], 397300B0-A, Version=2.1.2.1, Culture=neutral, PublicKeyToken=null]], 397300B0-B, Version=1.2.2.2, Culture=neutral, PublicKeyToken=null"),
+                    Add("1", "A`1[[B`1[[System.Object, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], 397300B0-B, Version=1.2.2.2, Culture=neutral, PublicKeyToken=null]], 397300B0-A, Version=2.1.2.1, Culture=neutral, PublicKeyToken=1f8a32457d187bf3").
+                    Add("2", "B`1[[A`1[[System.Object, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]][], 397300B0-A, Version=2.1.2.1, Culture=neutral, PublicKeyToken=1f8a32457d187bf3]], 397300B0-B, Version=1.2.2.2, Culture=neutral, PublicKeyToken=null"),
                 "(object)$exception ?? (object)$1 ?? $2",
                 DkmEvaluationFlags.TreatAsExpression,
                 DiagnosticFormatter.Instance,

--- a/src/ExpressionEvaluator/VisualBasic/Test/ExpressionCompiler/PseudoVariableTests.vb
+++ b/src/ExpressionEvaluator/VisualBasic/Test/ExpressionCompiler/PseudoVariableTests.vb
@@ -859,8 +859,8 @@ End Class"
             ' GetType(Exception), GetType(A(Of B(Of Object))), GetType(B(Of A(Of Object)()))
             context.CompileExpression(
                 InspectionContextFactory.Empty.Add("$exception", "System.Exception, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089").
-                    Add("1", "A`1[[B`1[[System.Object, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], 397300B1-B, Version=1.2.2.2, Culture=neutral, PublicKeyToken=null]], 397300B1-A, Version=2.1.2.1, Culture=neutral, PublicKeyToken=null").
-                    Add("2", "B`1[[A`1[[System.Object, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]][], 397300B1-A, Version=2.1.2.1, Culture=neutral, PublicKeyToken=null]], 397300B1-B, Version=1.2.2.2, Culture=neutral, PublicKeyToken=null"),
+                    Add("1", "A`1[[B`1[[System.Object, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], 397300B1-B, Version=1.2.2.2, Culture=neutral, PublicKeyToken=null]], 397300B1-A, Version=2.1.2.1, Culture=neutral, PublicKeyToken=1f8a32457d187bf3").
+                    Add("2", "B`1[[A`1[[System.Object, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]][], 397300B1-A, Version=2.1.2.1, Culture=neutral, PublicKeyToken=1f8a32457d187bf3]], 397300B1-B, Version=1.2.2.2, Culture=neutral, PublicKeyToken=null"),
                 "If(If($exception, $1), $2)",
                 DkmEvaluationFlags.TreatAsExpression,
                 DiagnosticFormatter.Instance,

